### PR TITLE
Integration spec

### PR DESCRIPTION
**Pull request by Wineshuga**
## Description
In this project, I and my partner @Simpleshaikh1 created integration tests for all of the views used in this project.

#### n+1
- The N+1 problem is solved when fetching all posts and their comments for a user.
**Before:**

first log

**After:**

2nd log

#### Integration specs

- We used Capybara to write integration tests for each view in the project.
- The following conditions were checked:
    - User index page:
        - I can see the username of all other users.
        - I can see the profile picture for each user.
        - I can see the number of posts each user has written.
        - When I click on a user, I am redirected to that user's show page.

    - user show page:
        - I can see the user's profile picture.
        - I can see the user's username.
        - I can see the number of posts the user has written.
        - I can see the user's bio.
        - I can see the user's first 3 posts.
        - I can see a button that lets me view all of a user's posts.
        - When I click a user's post, it redirects me to that post's show page.
        - When I click to see all posts, it redirects me to the user's post's index page.

    - User post index page:
        - I can see the user's profile picture.
        - I can see the user's username.
        - I can see the number of posts the user has written.
        - I can see a post's title.
        - I can see some of the post's body.
        - I can see the first comments on a post.
        - I can see how many comments a post has.
        - I can see how many likes a post has.
        - I can see a section for pagination if there are more posts than fit on the view.
        - When I click on a post, it redirects me to that post's show page.

    - Post-show page:
        - I can see the post's title.
        - I can see who wrote the post.
        - I can see how many comments it has.
        - I can see how many likes it has.
        - I can see the post body.
        - I can see the username of each commenter.
        - I can see the comment each commenter left.

### CODE REVIEW

File: app/views/posts/index.html.erb [hunk 1]
```diff
@@ -16,7 +16,7 @@
   <%= link_to "Add a new post", new_user_post_path(user_id: @user.id), class: "new-post-btn post-btn"%>
 
   <section class='post-container'>
-    <% @posts.each do |post| %>
+    <% @posts.includes(author: :comments).each do |post| %>
```
Comment (julie-ify) [inline app/views/posts/index.html.erb]: - [ ] The n+1 query problem should be eliminated in the controller action, as that handles the business logic of the app, including queries. The views should only be used for rendering data from the controller, it should not handle queries. See an example of how you may eliminate the problem in the controller.

  ```
  @posts = @user.posts.includes(:comments)
  ```

  image

File: app/views/posts/index.html.erb [hunk 2]
```diff
@@ -29,12 +29,15 @@
           </div>
           <h4 class="comment-title">Comments</h4>
           <div class="comment-container">
-            <% post.recent_comments.each do |comment|%>
+            <% post.comments.each do |comment|%>
```
Comment (julie-ify) [inline app/views/posts/index.html.erb]: - [ ] **[OPTIONAL]** If I recall correctly, from a previous project, the post index view should render the recent comments, not all comments. Kindly revert to the previous code you have.

File: spec/integration/user_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      expect(page).to have_content("Number of posts: #{@user.posts_counter}")
+    end
+
+    it "renders user's bio" do
+      expect(page).to have_content(@user.bio)
+    end
+
+    it 'renders the first 3 posts of the user' do
+      expect(page).to have_content(@post.title)
+    end
```
Comment (julie-ify) [inline spec/integration/user_show_spec.rb]: - [ ] This test requires that there should be up to 3 posts. For a more accurate assertion, consider creating more posts not just one.

File: spec/integration/user_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    it "redirects to post's index page" do
+      click_link 'See all posts'
+      expect(page).to have_current_path(user_posts_path(@user))
+    end
+
+    it "redirects to user's show page clicked" do
+      click_link 'See all posts'
+      expect(page).to have_current_path(user_posts_path(@user))
+    end
```
Comment (julie-ify) [inline spec/integration/user_show_spec.rb]: - [ ] These two tests are duplicates. The second one should check for a redirect to the post show page when a post is clicked. Kindly refactor it.

File: spec/integration/user_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+require 'rails_helper'
+
+RSpec.describe 'User Index Page', type: :system do
```
Comment (julie-ify) [inline spec/integration/user_index_spec.rb]: - [ ] All your tests in this file are failing. I think this has to do with the `type: :system` setup. When I changed it to `type: :feature` all the tests passed.

  image

**Review by julie-ify — CHANGES_REQUESTED** [review]
## Hi @Wineshuga and @Simpleshaikh1,

Good job so far!

## Highlights!
- Good code organization and passing integration tests ✔️
- Descriptive README ✔️ 

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@julie-ify` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Olamarx — APPROVED** [review]
Hi @Wineshuga,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- Linters passed ✔️
- Changes requested fixed ✔️
- Commit messages meaningful ✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**